workerpool/pool.go

```diff
@@ -12,75 +12,83 @@ import (
 	session     *concurrency.Session
+	events      <-chan *kv.Event
+	jobs        map[connection.JobID]*job // map of job handlers assigned to each connection.ID
 	log         *zap.SugaredLogger
-	jobs        map[connection.ID]*job // map of job handlers assigned to each connection.ID
-	events      <-chan *watcher.Event
 }
 
-// New will accept a few initializer variables in order to stand up the new worker
-// pool of goroutines. These workers will listen for *watcher.Events and handle the
-// internal *Connection to manage data.
-func New(session *concurrency.Session, maxWorkers uint, events <-chan *watcher.Event, locksPrefix string, log *zap.SugaredLogger) *WorkerPool {
+// Config is a configuration of the worker pool.
+type Config struct {
+	MaxWorkers  uint
+	LocksPrefix string
+	Session     *concurrency.Session
+	Events      <-chan *kv.Event
+	Log         *zap.SugaredLogger
+}
+
+// New creates and configures new worker pool. Worker pool has to be explicitly started with Start() method.
+func New(config *Config) *WorkerPool {
 	return &WorkerPool{
-		maxWorkers:  maxWorkers,
-		locksPrefix: locksPrefix,
-		session:     session,
-		log:         log,
-		jobs:        make(map[connection.ID]*job),
-		events:      events,
+		maxWorkers:  config.MaxWorkers,
+		locksPrefix: config.LocksPrefix,
+		session:     config.Session,
+		events:      config.Events,
+		jobs:        make(map[connection.JobID]*job),
+		log:         config.Log,
 	}
 }
 
-// Start listens on events channel and tries to start a job for every connection.
+// Start listens on kv.Event channel, tries to create a lock, and start a job.
 func (pool *WorkerPool) Start() {
 	go func() {
 		for {
 			event, more := <-pool.events
 			if more {
-				pool.log.Debugw("event received", "connectionID", event.ID, "type", event.Type)
+				pool.log.Debugw("event received", "jobID", event.JobID, "type", event.Type)
 
 				switch event.Type {
-				case watcher.Created:
-					if pool.numWorkers+event.Connection.Source.NumberOfWorkers() > pool.maxWorkers {
-						pool.log.Debugw("creating new job skipped, workers limit exceeded", "connectionID", event.ID)
+				case kv.Created:
+					if pool.numWorkers+event.Job.NumberOfWorkers > pool.maxWorkers {
+						pool.log.Debugw("creating new job skipped, workers limit exceeded", "jobID", event.JobID)
 						continue
 					}
 
-					job, err := newJob(pool.session, event.Connection, pool.locksPrefix, pool.log.Named("job"))
+					job, err := newJob(pool.session, event.Job, pool.locksPrefix, pool.log.Named("job"))
 					if err != nil {
-						pool.log.Debugw("creating new job failed", "error", err, "connectionID", event.ID)
+						pool.log.Debugw("creating new job failed", "error", err, "jobID", event.JobID)
 						continue
 					}
 
 					job.start()
 
-					pool.jobs[event.ID] = job
-					pool.numWorkers += event.Connection.Source.NumberOfWorkers()
-				case watcher.Deleted:
-					if job, exists := pool.jobs[event.ID]; exists {
+					pool.jobs[event.JobID] = job
```

Comment:
Is there any risk of concurrent map read/writes if we fork this into its own goroutine? If the job gets killed or cancelled or for any other reason ends up in `Stop` state while a new job enters, I think this would panic.

I've had good success with the `sync.Map` for concurrent use-cases, might want to think about that here too.

Reply:
yeah, or just use RWMutex.